<a href="https://colab.research.google.com/github/womaro/Story/blob/main/Seaborn_and_Linear_Regression_(start).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

Do higher film budgets lead to more box office revenue? Let's find out if there's a relationship using the movie budgets and financial performance data that I've scraped from [the-numbers.com](https://www.the-numbers.com/movie/budgets) on **May 1st, 2018**. 

<img src=https://i.imgur.com/kq7hrEh.png>

# Import Statements

In [1]:
import pandas as pd
import matplotlib.pyplot as plt


# Notebook Presentation

In [2]:
pd.options.display.float_format = '{:,.2f}'.format

from pandas.plotting import register_matplotlib_converters
register_matplotlib_converters()

# Read the Data

In [3]:
data = pd.read_csv('cost_revenue_dirty.csv')

# Explore and Clean the Data

**Challenge**: Answer these questions about the dataset:
1. How many rows and columns does the dataset contain?
2. Are there any NaN values present?
3. Are there any duplicate rows?
4. What are the data types of the columns?

In [4]:
data.shape

(5391, 6)

In [12]:
data.sample(10).sort_values('USD_Production_Budget', ascending=False)

Rank Release_Date                   Movie_Title USD_Production_Budget  \
1894  3756   10/18/2002                    Auto Focus            $7,000,000   
1269   731    7/16/1999                Eyes Wide Shut           $65,000,000   
2010   996    4/16/2003              Bulletproof Monk           $52,000,000   
1857  1046     8/5/2002                    Unfaithful           $50,000,000   
5300  1452   12/23/2016                      Why Him?           $38,000,000   
992   4266    8/13/1997                The Full Monty            $3,500,000   
1479  4336    8/18/2000  The Original Kings of Comedy            $3,000,000   
3177   149    7/11/2008   Madagascar: Escape 2 Africa          $150,000,000   
1018  5372    11/4/1997                Pink Flamingos               $12,000   
4218  4816    8/28/2012                    Below Zero            $1,200,000   

     USD_Worldwide_Gross USD_Domestic_Gross  
1894          $2,703,821         $2,062,066  
1269        $104,267,443        $55,691,208  
2010         $23,010,607        $23,010,607  
1857        $119,114,494        $52,752,475  
5300        $117,523,786        $60,323,786  
992         $257,938,649        $45,950,122  
1479         $38,236,338        $38,168,022  
3177        $599,516,844       $180,010,950  
1018            $413,802           $413,802  
4218                  $0                 $0

In [13]:
data.isna().sum()

Rank                     0
Release_Date             0
Movie_Title              0
USD_Production_Budget    0
USD_Worldwide_Gross      0
USD_Domestic_Gross       0
dtype: int64

In [15]:
duplicates = data[data.duplicated()]
duplicates.head()

Empty DataFrame
Columns: [Rank, Release_Date, Movie_Title, USD_Production_Budget, USD_Worldwide_Gross, USD_Domestic_Gross]
Index: []

In [16]:
duplicates = data[data.duplicated(subset=['Movie_Title'])]
duplicates.head()

Rank Release_Date                   Movie_Title USD_Production_Budget  \
64   3973   12/23/1954  20,000 Leagues Under the Sea            $5,000,000   
84   2802   11/18/1959                       Ben-Hur           $15,000,000   
215  2202   12/17/1976                     King Kong           $23,000,000   
322  3105     2/4/1982                    Cat People           $12,500,000   
579  4154   10/19/1990      Night of the Living Dead            $4,200,000   

    USD_Worldwide_Gross USD_Domestic_Gross  
64          $28,200,000        $28,200,000  
84          $73,000,000        $73,000,000  
215         $90,614,445        $52,614,445  
322         $21,000,000         $7,000,000  
579          $5,835,247         $5,835,247

In [17]:
data[data.Movie_Title=='Ben-Hur']

Rank Release_Date Movie_Title USD_Production_Budget USD_Worldwide_Gross  \
5     4250   12/30/1925     Ben-Hur            $3,900,000          $9,000,000   
84    2802   11/18/1959     Ben-Hur           $15,000,000         $73,000,000   
5231   380    8/19/2016     Ben-Hur           $95,000,000         $88,725,100   

     USD_Domestic_Gross  
5            $9,000,000  
84          $73,000,000  
5231        $26,410,477

In [18]:
duplicates = data[data.duplicated(subset=['Movie_Title', 'Release_Date'])]
duplicates.head()

Rank Release_Date        Movie_Title USD_Production_Budget  \
3055  5290   12/31/2007  A Dog's Breakfast              $120,000   

     USD_Worldwide_Gross USD_Domestic_Gross  
3055                  $0                 $0

In [19]:
data[data.Movie_Title=="A Dog's Breakfast"]

Rank Release_Date        Movie_Title USD_Production_Budget  \
3054  5289   12/31/2007  A Dog's Breakfast              $120,000   
3055  5290   12/31/2007  A Dog's Breakfast              $120,000   

     USD_Worldwide_Gross USD_Domestic_Gross  
3054                  $0                 $0  
3055                  $0                 $0

In [20]:
data = data.drop_duplicates(subset=['Movie_Title', 'Release_Date'])

In [21]:
duplicates = data[data.duplicated(subset=['Movie_Title', 'Release_Date'])]
duplicates.head()

Empty DataFrame
Columns: [Rank, Release_Date, Movie_Title, USD_Production_Budget, USD_Worldwide_Gross, USD_Domestic_Gross]
Index: []

In [23]:
data.describe()

Rank
count 5,390.00
mean  2,695.52
std   1,556.14
min       1.00
25%   1,348.25
50%   2,695.50
75%   4,042.75
max   5,391.00

In [24]:
data.dtypes

Rank                      int64
Release_Date             object
Movie_Title              object
USD_Production_Budget    object
USD_Worldwide_Gross      object
USD_Domestic_Gross       object
dtype: object

### Data Type Conversions

**Challenge**: Convert the `USD_Production_Budget`, `USD_Worldwide_Gross`, and `USD_Domestic_Gross` columns to a numeric format by removing `$` signs and `,`. 
<br>
<br>
Note that *domestic* in this context refers to the United States.

In [31]:
data.USD_Production_Budget = data.USD_Production_Budget.astype(str).str.replace('$', '').str.replace(',','')
data.USD_Worldwide_Gross = data.USD_Worldwide_Gross.astype(str).str.replace('$', '').str.replace(',','')
data.USD_Domestic_Gross = data.USD_Domestic_Gross.astype(str).str.replace('$', '').str.replace(',','')

data.USD_Production_Budget = pd.to_numeric(data.USD_Production_Budget)
data.USD_Worldwide_Gross = pd.to_numeric(data.USD_Worldwide_Gross)
data.USD_Domestic_Gross = pd.to_numeric(data.USD_Domestic_Gross)

data.dtypes

<ipython-input-31-e97f671f82c1>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data.USD_Production_Budget = data.USD_Production_Budget.astype(str).str.replace('$', '').str.replace(',','')
<ipython-input-31-e97f671f82c1>:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data.USD_Worldwide_Gross = data.USD_Worldwide_Gross.astype(str).str.replace('$', '').str.replace(',','')
<ipython-input-31-e97f671f82c1>:3: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  data.USD_Domestic_Gross = data.USD_Domestic_Gross.astype(str).str.rep

Rank                      int64
Release_Date             object
Movie_Title              object
USD_Production_Budget     int64
USD_Worldwide_Gross       int64
USD_Domestic_Gross        int64
dtype: object

**Challenge**: Convert the `Release_Date` column to a Pandas Datetime type. 

In [36]:
data.Release_Date = pd.to_datetime(data.Release_Date)
data.dtypes

Rank                              int64
Release_Date             datetime64[ns]
Movie_Title                      object
USD_Production_Budget             int64
USD_Worldwide_Gross               int64
USD_Domestic_Gross                int64
dtype: object

In [37]:
data.sample(3)

Rank Release_Date                   Movie_Title  USD_Production_Budget  \
4522  1860   2013-12-31  Space Pirate Captain Harlock               30000000   
1917  5149   2002-11-15                  Revolution#9                 350000   
4041   464   2011-11-23              Arthur Christmas               85000000   

      USD_Worldwide_Gross  USD_Domestic_Gross  
4522               143490                   0  
1917                 9118                9118  
4041            149717124            46462469

### Descriptive Statistics

**Challenge**: 

1. What is the average production budget of the films in the data set?
2. What is the average worldwide gross revenue of films?
3. What were the minimums for worldwide and domestic revenue?
4. Are the bottom 25% of films actually profitable or do they lose money?
5. What are the highest production budget and highest worldwide gross revenue of any film?
6. How much revenue did the lowest and highest budget films make?

In [33]:
data.describe()

Rank  USD_Production_Budget  USD_Worldwide_Gross  USD_Domestic_Gross
count 5,390.00               5,390.00             5,390.00            5,390.00
mean  2,695.52          31,119,487.81        88,871,907.20       41,243,169.82
std   1,556.14          40,525,356.93       168,469,037.10       66,033,082.74
min       1.00               1,100.00                 0.00                0.00
25%   1,348.25           5,000,000.00         3,877,146.00        1,335,065.75
50%   2,695.50          17,000,000.00        27,460,037.00       17,196,565.00
75%   4,042.75          40,000,000.00        96,461,821.00       52,348,661.50
max   5,391.00         425,000,000.00     2,783,918,982.00      936,662,225.00

# Investigating the Zero Revenue Films

**Challenge** How many films grossed $0 domestically (i.e., in the United States)? What were the highest budget films that grossed nothing?

In [38]:
data[data.USD_Domestic_Gross==0].count()

Rank                     511
Release_Date             511
Movie_Title              511
USD_Production_Budget    511
USD_Worldwide_Gross      511
USD_Domestic_Gross       511
dtype: int64

In [39]:
data[data.USD_Domestic_Gross==0].sort_values('USD_Production_Budget', ascending=False).head(5)

Rank Release_Date                         Movie_Title  \
5388    96   2020-12-31                         Singularity   
5387   126   2018-12-18                             Aquaman   
5384   321   2018-09-03                   A Wrinkle in Time   
5385   366   2018-10-08                      Amusement Park   
5090   556   2015-12-31  Don Gato, el inicio de la pandilla   

      USD_Production_Budget  USD_Worldwide_Gross  USD_Domestic_Gross  
5388              175000000                    0                   0  
5387              160000000                    0                   0  
5384              103000000                    0                   0  
5385              100000000                    0                   0  
5090               80000000              4547660                   0

**Challenge**: How many films grossed $0 worldwide? What are the highest budget films that had no revenue internationally?

In [40]:
data[data.USD_Worldwide_Gross==0].count()

Rank                     356
Release_Date             356
Movie_Title              356
USD_Production_Budget    356
USD_Worldwide_Gross      356
USD_Domestic_Gross       356
dtype: int64

In [42]:
data[data.USD_Worldwide_Gross==0].sort_values('USD_Production_Budget', ascending=False).head(5)

Rank Release_Date        Movie_Title  USD_Production_Budget  \
5388    96   2020-12-31        Singularity              175000000   
5387   126   2018-12-18            Aquaman              160000000   
5384   321   2018-09-03  A Wrinkle in Time              103000000   
5385   366   2018-10-08     Amusement Park              100000000   
5338   879   2017-04-08     The Dark Tower               60000000   

      USD_Worldwide_Gross  USD_Domestic_Gross  
5388                    0                   0  
5387                    0                   0  
5384                    0                   0  
5385                    0                   0  
5338                    0                   0

### Filtering on Multiple Conditions

In [44]:
international_releases = data.loc[(data.USD_Domestic_Gross == 0) & (data.USD_Worldwide_Gross != 0)]
international_releases.head()

Rank Release_Date              Movie_Title  USD_Production_Budget  \
71    4310   1956-02-16                 Carousel                3380000   
1579  5087   2001-02-11  Everything Put Together                 500000   
1744  3695   2001-12-31                 The Hole                7500000   
2155  4236   2003-12-31                  Nothing                4000000   
2203  2513   2004-03-31                The Touch               20000000   

      USD_Worldwide_Gross  USD_Domestic_Gross  
71                   3220                   0  
1579                 7890                   0  
1744             10834406                   0  
2155                63180                   0  
2203              5918742                   0

**Challenge**: Use the [`.query()` function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.query.html) to accomplish the same thing. Create a subset for international releases that had some worldwide gross revenue, but made zero revenue in the United States. 

Hint: This time you'll have to use the `and` keyword.

### Unreleased Films

**Challenge**:
* Identify which films were not released yet as of the time of data collection (May 1st, 2018).
* How many films are included in the dataset that have not yet had a chance to be screened in the box office? 
* Create another DataFrame called data_clean that does not include these films. 

In [ ]:
# Date of Data Collection
scrape_date = pd.Timestamp('2018-5-1')

### Films that Lost Money

**Challenge**: 
What is the percentage of films where the production costs exceeded the worldwide gross revenue? 

# Seaborn for Data Viz: Bubble Charts

### Plotting Movie Releases over Time

**Challenge**: Try to create the following Bubble Chart:

<img src=https://i.imgur.com/8fUn9T6.png>



# Converting Years to Decades Trick

**Challenge**: Create a column in `data_clean` that has the decade of the release. 

<img src=https://i.imgur.com/0VEfagw.png width=650> 

Here's how: 
1. Create a [`DatetimeIndex` object](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DatetimeIndex.html) from the Release_Date column. 
2. Grab all the years from the `DatetimeIndex` object using the `.year` property.
<img src=https://i.imgur.com/5m06Ach.png width=650>
3. Use floor division `//` to convert the year data to the decades of the films.
4. Add the decades as a `Decade` column to the `data_clean` DataFrame.

### Separate the "old" (before 1969) and "New" (1970s onwards) Films

**Challenge**: Create two new DataFrames: `old_films` and `new_films`
* `old_films` should include all the films before 1969 (up to and including 1969)
* `new_films` should include all the films from 1970 onwards
* How many films were released prior to 1970?
* What was the most expensive film made prior to 1970?

# Seaborn Regression Plots

**Challenge**: Use Seaborn's `.regplot()` to show the scatter plot and linear regression line against the `new_films`. 
<br>
<br>
Style the chart

* Put the chart on a `'darkgrid'`.
* Set limits on the axes so that they don't show negative values.
* Label the axes on the plot "Revenue in \$ billions" and "Budget in \$ millions".
* Provide HEX colour codes for the plot and the regression line. Make the dots dark blue (#2f4b7c) and the line orange (#ff7c43).

Interpret the chart

* Do our data points for the new films align better or worse with the linear regression than for our older films?
* Roughly how much would a film with a budget of $150 million make according to the regression line?

# Run Your Own Regression with scikit-learn

$$ REV \hat ENUE = \theta _0 + \theta _1 BUDGET$$

**Challenge**: Run a linear regression for the `old_films`. Calculate the intercept, slope and r-squared. How much of the variance in movie revenue does the linear model explain in this case?

# Use Your Model to Make a Prediction

We just estimated the slope and intercept! Remember that our Linear Model has the following form:

$$ REV \hat ENUE = \theta _0 + \theta _1 BUDGET$$

**Challenge**:  How much global revenue does our model estimate for a film with a budget of $350 million? 